lib/multiplexer/proxyline.go

```diff
@@ -33,27 +35,49 @@ import (
 	"github.com/gravitational/trace"
 )
 
+// PP2Type is the PROXY protocol v2 TLV type
+type PP2Type byte
+
 const (
 	// TCP4 is TCP over IPv4
 	TCP4 = "TCP4"
 	// TCP6 is tCP over IPv6
 	TCP6 = "TCP6"
 	// Unknown is unsupported or unknown protocol
 	UNKNOWN = "UNKNOWN"
+
+	PP2TypeNOOP PP2Type = 0x04 // No-op used for padding
+
+	// Known custom types, spec allows to use 0xE0 - 0xEF for custom types
+	PP2TypeGCP   PP2Type = 0xE0 // https://cloud.google.com/vpc/docs/configure-private-service-connect-producer
+	PP2TypeAWS   PP2Type = 0xEA // https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-target-groups.html
+	PP2TypeAzure PP2Type = 0xEE // https://learn.microsoft.com/en-us/azure/private-link/private-link-service-overview
+
+	PP2TypeTeleport PP2Type = 0xE4 // Teleport own type for transferring our custom data
```

Comment:
Maybe we ought to be a little more specific as to what "custom data" is. As the RFD is private, I think we could do with making sure this implementation is very clear within the source as to what it does.

I'd suggest `// Teleport's type for transferring signed connection metadata from the Proxy to Auth servers and Nodes`

Reply:
Technically this is not linked directly to RFD and can be used for any data, it's intended just for general usage (one of cases of which will be from the RFD). I think we should do like AWS - they have top level `PP2_TYPE_AWS` and then subtype `PP2_SUBTYPE_AWS_VPCE_ID`. So we can transfer metadata for the RFD as subtypes (basically TLV inside TLV) - that way we are more free (fx if we want to introduce new subtypes) we don't need to fiddle with top level types, where space is kinda limited. I didn't want to add subtypes in this PR since they don't make much sense without everything else.

Reply:
Ah - if there's a plan for some subtyping or similar, I think that may make sense then.

I would probably still extend the comment a little bit, I think we don't do a great job across the Teleport codebase of signposting what things are often used for, and without the context of why a piece of code exists it can be really hard to work out if something has a correct implementation.

`// Teleport's own type for transferring custom data such as connection metadata` ? I just feel like "custom data' could really mean anything to somebody trying to understand `multiplexer`.

Reply:
@strideynet yep, will extend the comment 👍